qualtran/bloqs/arithmetic/addition.py

```diff
@@ -162,29 +191,30 @@ def decompose_from_registers(
         # reverse the order of qubits for big endian-ness.
         input_bits = quregs['a'][::-1]
         output_bits = quregs['b'][::-1]
-        ancillas = context.qubit_manager.qalloc(self.dtype.bitsize - 1)[::-1]
+        ancillas = context.qubit_manager.qalloc(self.b_dtype.bitsize - 1)[::-1]
         # Start off the addition by anding into the ancilla
         yield And().on(input_bits[0], output_bits[0], ancillas[0])
         # Left part of Fig.2
         yield from self._left_building_block(input_bits, output_bits, ancillas, 1)
         yield CNOT().on(ancillas[-1], output_bits[-1])
-        yield CNOT().on(input_bits[-1], output_bits[-1])
+        if len(input_bits) == len(output_bits):
+            yield CNOT().on(input_bits[-1], output_bits[-1])
         # right part of Fig.2
         yield from self._right_building_block(
-            input_bits, output_bits, ancillas, self.dtype.bitsize - 2
+            input_bits, output_bits, ancillas, self.b_dtype.bitsize - 2
         )
         yield And().adjoint().on(input_bits[0], output_bits[0], ancillas[0])
         yield CNOT().on(input_bits[0], output_bits[0])
         context.qubit_manager.qfree(ancillas)
 
     def _t_complexity_(self):
-        n = self.dtype.bitsize
+        n = self.b_dtype.bitsize
         num_clifford = (n - 2) * 19 + 16
         num_toffoli = n - 1
         return TComplexity(t=4 * num_toffoli, clifford=num_clifford)
 
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
-        n = self.dtype.bitsize
+        n = self.b_dtype.bitsize
```

Comment:
why are we using the second registers' bitsize?

Reply:
This the register that we are adding into and is always >= first register size.